config/replacements.php

```diff
@@ -151,112 +151,138 @@
-    'm\\ZendOAuth\\' => 'm\\ZendOAuth\\',
-    'n\\ZendOAuth\\' => 'n\\ZendOAuth\\',
-    'o\\ZendOAuth\\' => 'o\\ZendOAuth\\',
-    'p\\ZendOAuth\\' => 'p\\ZendOAuth\\',
-    'q\\ZendOAuth\\' => 'q\\ZendOAuth\\',
-    'r\\ZendOAuth\\' => 'r\\ZendOAuth\\',
-    's\\ZendOAuth\\' => 's\\ZendOAuth\\',
-    't\\ZendOAuth\\' => 't\\ZendOAuth\\',
-    'u\\ZendOAuth\\' => 'u\\ZendOAuth\\',
-    'v\\ZendOAuth\\' => 'v\\ZendOAuth\\',
-    'w\\ZendOAuth\\' => 'w\\ZendOAuth\\',
-    'x\\ZendOAuth\\' => 'x\\ZendOAuth\\',
-    'y\\ZendOAuth\\' => 'y\\ZendOAuth\\',
-    'z\\ZendOAuth\\' => 'z\\ZendOAuth\\',
-    'a\\ZendService\\' => 'a\\ZendService\\',
-    'b\\ZendService\\' => 'b\\ZendService\\',
-    'c\\ZendService\\' => 'c\\ZendService\\',
-    'd\\ZendService\\' => 'd\\ZendService\\',
-    'e\\ZendService\\' => 'e\\ZendService\\',
-    'f\\ZendService\\' => 'f\\ZendService\\',
-    'g\\ZendService\\' => 'g\\ZendService\\',
-    'h\\ZendService\\' => 'h\\ZendService\\',
-    'i\\ZendService\\' => 'i\\ZendService\\',
-    'j\\ZendService\\' => 'j\\ZendService\\',
-    'k\\ZendService\\' => 'k\\ZendService\\',
-    'l\\ZendService\\' => 'l\\ZendService\\',
-    'm\\ZendService\\' => 'm\\ZendService\\',
-    'n\\ZendService\\' => 'n\\ZendService\\',
-    'o\\ZendService\\' => 'o\\ZendService\\',
-    'p\\ZendService\\' => 'p\\ZendService\\',
-    'q\\ZendService\\' => 'q\\ZendService\\',
-    'r\\ZendService\\' => 'r\\ZendService\\',
-    's\\ZendService\\' => 's\\ZendService\\',
-    't\\ZendService\\' => 't\\ZendService\\',
-    'u\\ZendService\\' => 'u\\ZendService\\',
-    'v\\ZendService\\' => 'v\\ZendService\\',
-    'w\\ZendService\\' => 'w\\ZendService\\',
-    'x\\ZendService\\' => 'x\\ZendService\\',
-    'y\\ZendService\\' => 'y\\ZendService\\',
-    'z\\ZendService\\' => 'z\\ZendService\\',
-    'a\\ZendXml\\' => 'a\\ZendXml\\',
-    'b\\ZendXml\\' => 'b\\ZendXml\\',
-    'c\\ZendXml\\' => 'c\\ZendXml\\',
-    'd\\ZendXml\\' => 'd\\ZendXml\\',
-    'e\\ZendXml\\' => 'e\\ZendXml\\',
-    'f\\ZendXml\\' => 'f\\ZendXml\\',
-    'g\\ZendXml\\' => 'g\\ZendXml\\',
-    'h\\ZendXml\\' => 'h\\ZendXml\\',
-    'i\\ZendXml\\' => 'i\\ZendXml\\',
-    'j\\ZendXml\\' => 'j\\ZendXml\\',
-    'k\\ZendXml\\' => 'k\\ZendXml\\',
-    'l\\ZendXml\\' => 'l\\ZendXml\\',
-    'm\\ZendXml\\' => 'm\\ZendXml\\',
-    'n\\ZendXml\\' => 'n\\ZendXml\\',
-    'o\\ZendXml\\' => 'o\\ZendXml\\',
-    'p\\ZendXml\\' => 'p\\ZendXml\\',
-    'q\\ZendXml\\' => 'q\\ZendXml\\',
-    'r\\ZendXml\\' => 'r\\ZendXml\\',
-    's\\ZendXml\\' => 's\\ZendXml\\',
-    't\\ZendXml\\' => 't\\ZendXml\\',
-    'u\\ZendXml\\' => 'u\\ZendXml\\',
-    'v\\ZendXml\\' => 'v\\ZendXml\\',
-    'w\\ZendXml\\' => 'w\\ZendXml\\',
-    'x\\ZendXml\\' => 'x\\ZendXml\\',
-    'y\\ZendXml\\' => 'y\\ZendXml\\',
-    'z\\ZendXml\\' => 'z\\ZendXml\\',
+    // class ZendAcl in zend-expressive-authorization-acl
+    'ZendAcl' => 'LaminasAcl',
+    'Acl\\ZendAcl' => 'Acl\\LaminasAcl',
```

Comment:
Do we have test case:
```php
'factories' => [
    'any-alias' => \Some\Vendor\Acl\ZendAcl::class,
],
```
?

Reply:
I'll add one.

Reply:
Fails. And that puts us in a strange area, because we really cannot control _every_ possible context.

I think we can only do:

- Bare name (e.g., `ZendAcl` becomes `LaminasAcl`)
- FCQN (e.g., `Zend\Expressive\Authorization\Acl\ZendAcl` becomes `Mezzio\Authorization\Acl\LaminasAcl`)

If we do the parent namespace as well, we start running into issues with other libraries too quickly.

I've just tried that approach (removing the version with the parent subnamespace), and all tests pass, including the new one you suggested, so I'm going to consider that a win. I'll push shortly.

Reply:
And in that case migration will fail on something like that:

```php
use Zend\Expressive\Authorization\Acl;

Acl\ZendAcl::class;
```

it will be migrated to:

```php
use Mezzio\Authorization\Acl;

Acl\ZendAcl::class;
```

we should be aware of it - maybe add it into FAQ of the migration?

Reply:
Also another case:
```php
use Some\Vendor\ZendAcl;

ZendAcl::class;
```

after migration it becomes:

```php
use Some\Vendor\ZendAcl;

LaminasAcl::class; // <-- wrong as this class does not exist, it should be ZendAcl::class
```

As in the previous comment, probably we cannot cover all cases so we should only add it into FAQ or Migration notes/issues.

More important is that PostProcessor replaces classes correctly and I believe we have achieved it here.

Reply:
Agreed - these will need to be part of the FAQ.

I don't think they'll happen often, but they _can_ happen. I added test cases, and verified that they fail, and then removed them, as we expect those failures currently.

Reply:
@weierophinney created issue for that: https://github.com/laminas/laminas.github.io/issues/63